src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -508,7 +549,47 @@ public T GetHeaderValue<T>(string key)
             if (string.IsNullOrEmpty(key))
                 throw LogHelper.LogArgumentNullException(nameof(key));
 
-            return Header.Value<T>(key);
+            var result = Header.TryGetValue(key, out var jTokenValue);
+            if (jTokenValue == null)
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14303, key)));
+
+            T value;
+            try
+            {
+                value =  jTokenValue.ToObject<T>();
+            }
+            catch (Exception ex)
+            {
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14305, key, typeof(T), jTokenValue.Type), ex));
+            }
+
+            return value;
+        }
+
+        /// <summary>
+        /// Tries to get the value corresponding to the provided key from the JWT header { key, 'value' }.
+        /// </summary>
+        /// <remarks>If the key has no corresponding value, returns false. Otherwise returns true. </remarks>   
+        public bool TryGetHeaderValue<T>(string key, out T value)
+        {
+            if (string.IsNullOrEmpty(key))
+            {
+                value = default(T);
+                return false;
+            }
+
+            var result = Header.TryGetValue(key, out var jTokenValue);
+            try
```

Comment:
if 'result' is false, then do we need to execute the try  - catch?
we are making the assumption that all call graphs will set jTokenValue to null, can we assume this?
we could write:
```
if (!Header.TryGetValue ...)
{ 
	value = default(T);
	return false;
}

try
{
	value = jTokenValue.ToObject<T>();
}

Reply:
Good suggestion! 👍 You're right; the 'result' variable is unnecessary here. I've made similar changes to the other methods as well.